### PR TITLE
Added API to get garbage collection data from data stores and DDSs

### LINKED ISSUE #4533: Add a new API to get GC data from Data stores and DDS
Currently, GC data is sent as part of summary. It should be separated into its own API so that it can be called before summarize.

### PR DESCRIPTION
**Pull request by agarwal-navin**
Fixes #4533.

Added a new API to get GC data. Currently, GC nodes are returned as part of summary result. It will be removed in a follow up change to this. We need two separate APIs because summary result needs to include state about whether the node that is summarizing is referenced or not.
Running summary on the Container will have the following steps:
1. Use the new API to get GC data.
2. Run GC on the GC data.
3. Mark Fluid objects are referenced / unreferenced based on GC data.
4. Summarize the Fluid objects. The result will include state indicating whether it is referenced or not.

Also, updated GC nodes to be a map instead of an array in the new API. This has couple of advantages:
- Its easier to compare summaries that have GC nodes by deep comparing the two objects. With arrays, the arrays have to be sorted first.
- When running GC on the nodes, we don't need to convert the array to map for routes lookup.

### CODE REVIEW

File: packages/dds/shared-object-base/src/sharedObject.ts
Comment (vladsud) [inline packages/dds/shared-object-base/src/sharedObject.ts:242]: Maybe put a comment above _isSummarizing  on why we are doing so?

File: packages/dds/shared-object-base/src/sharedObject.ts
Comment (vladsud) [inline packages/dds/shared-object-base/src/sharedObject.ts:248]: I think it's better to reference a future issue that will track the work to make it more efficient and reference it here.
Also we need to have some readme.md or something explaining the overall flow - it might be hard for uneducated user looking at your PR (or code after it lands) to reason about overall flow, why it's done this way, etc.

File: packages/dds/shared-object-base/src/sharedObject.ts
Comment (vladsud) [inline packages/dds/shared-object-base/src/sharedObject.ts:251]: I'd add assert above it that it was still true before we reset. Similar (but reverse) at beginning of function - these asserts will catch reentrancy if it were to happen.

File: packages/runtime/container-runtime/src/dataStores.ts
Comment (vladsud) [inline packages/runtime/container-runtime/src/dataStores.ts:451]: I'm not 100% sure this assert is correct. Did you want to do != AttachState.Detached ?
Also should it be a check, not assert? I.e. it's totally normal to have unattached data stores

File: packages/runtime/datastore/src/dataStoreRuntime.ts
Comment (vladsud) [inline packages/runtime/datastore/src/dataStoreRuntime.ts:626]: Similar comment here - should this be a runtime check instead of assert?

File: packages/runtime/runtime-definitions/src/garbageCollection.ts
Comment (vladsud) [inline packages/runtime/runtime-definitions/src/garbageCollection.ts:12]: This looks like a Map<string, string[]>

File: packages/runtime/runtime-utils/src/summarizerNode/summarizerNodeWithGc.ts
Comment (vladsud) [inline packages/runtime/runtime-utils/src/summarizerNode/summarizerNodeWithGc.ts:65]: I've sent you an IM with more thoughts on this. But I believe it has to return a set of actual nodes that are left after GC was run. This info needs to be used on future GCs/summarization runs if this data store is not going to be loaded due to delay-loading.

**Review by vladsud — APPROVED** [review]
FYI - I only scanned through it briefly - would be useful if you get another pair of eyes on it.

File: packages/runtime/container-runtime/src/dataStores.ts
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/dataStores.ts:451]: I borrowed this logic from summarize. I reckon its because summarize / GC should only run in summarizer client which cannot have local contexts / channels.
Now that I think about it, it does not seem correct because summarize is a public API and it can be called on clients with local state too. Also, snapshot API in Container eventually calls this which might be problem if called on a Container that has local state.

Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/dataStores.ts:451]: But what happens if a data store / channel context is detached or attaching? Do we just ignore it for snapshot purposes? That also does not seem right.

Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/dataStores.ts:451]: For GC, not including detached stores / channels makes sense because they are not connected to the graph. So any references it might have are not connected to rest of the system.

File: packages/dds/shared-object-base/src/sharedObject.ts
Comment (agarwal-navin) [reply packages/dds/shared-object-base/src/sharedObject.ts:248]: Created an issue and referenced it here.
Yeah, a README would be good. Both for overall GC and how getting outbound routes work in DDSs. Created an issue #4548 to track this.

File: packages/runtime/runtime-definitions/src/garbageCollection.ts
Comment (agarwal-navin) [reply packages/runtime/runtime-definitions/src/garbageCollection.ts:12]: I like object more than a map here simply because it works with JSON. This will be useful when writing this to summary. For map, we will have to convert to an object to use with JSON.stringify.

File: packages/runtime/runtime-utils/src/summarizerNode/summarizerNodeWithGc.ts
Comment (agarwal-navin) [reply packages/runtime/runtime-utils/src/summarizerNode/summarizerNodeWithGc.ts]: As per our discussion, I will keep gcNodes in summarize for now. When we implement GC at DDS / sub-DDS level, we can revisit this and decide then if we want to remove it.

I have added the following commit that updates summarize to return IGCData instead of gcNodes to make it consistent with getGCData API:
https://github.com/microsoft/FluidFramework/pull/4538/commits/63c52bcc9c01ca2cd4dccfcc30570dd3eac1cc40
I wanted to show this in a separate PR but that seems to be rather difficult without merging this first :-)

**Comment by agarwal-navin** [comment]
Added a separate commit that updates summarize to return IGCData instead of gcNodes. This makes it consistent with getGCData API.
Here's the commit in case looking at it separately helps: https://github.com/microsoft/FluidFramework/commit/63c52bcc9c01ca2cd4dccfcc30570dd3eac1cc40

File: packages/dds/shared-object-base/src/sharedObject.ts
Comment (anthony-murphy) [inline packages/dds/shared-object-base/src/sharedObject.ts:228]: this should go in a finally, otherwise if an exception is thrown the object will always have this._isSummarizing = true;

File: packages/dds/shared-object-base/src/sharedObject.ts
Comment (anthony-murphy) [inline packages/dds/shared-object-base/src/sharedObject.ts:255]: >_isSummarizing [](start = 13, length = 14)

same as above

**Review by anthony-murphy — APPROVED** [review]
:shipit:

**Pull request merged**
